src/Mod/Sketcher/Gui/SnapManager.cpp

```diff
@@ -0,0 +1,309 @@
+
+        const Part::Geometry* geo = Obj->getGeometry(CrvId);
+
+        Base::Vector3d pointToOverride(x, y, 0.);
+
+        double pointParam;
+        auto curve = dynamic_cast<const Part::GeomCurve*>(geo);
+        if (curve != nullptr) {
+            try {
+                curve->closestParameter(pointToOverride, pointParam);
+                pointToOverride = curve->pointAtParameter(pointParam);
+            }
+            catch (Base::CADKernelError& e) {
+                e.ReportException();
+                return false;
+            }
+
+            //If it is a line, then we check if we need to snap to the middle.
+            if (geo->getTypeId() == Part::GeomLineSegment::getClassTypeId()) {
+                const Part::GeomLineSegment* line = static_cast<const Part::GeomLineSegment*>(geo);
+                snapToLineMiddle(pointToOverride, line);
+            }
+
+            //If it is an arc, then we check if we need to snap to the middle (not the center).
+            if (geo->getTypeId() == Part::GeomArcOfCircle::getClassTypeId()) {
+                const Part::GeomArcOfCircle* arc = static_cast<const Part::GeomArcOfCircle*>(geo);
+                snapToArcMiddle(pointToOverride, arc);
+            }
+
+            x = pointToOverride.x;
+            y = pointToOverride.y;
+
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void SnapManager::snapToGrid(double& x, double& y)
+{
+    // Snap Tolerance in pixels
+    const double snapTol = viewProvider.GridSize.getValue() / 5;
+
+    double tmpX = x, tmpY = y;
+
+    // Find Nearest Snap points
+    tmpX = tmpX / viewProvider.GridSize.getValue();
+    tmpX = tmpX < 0.0 ? ceil(tmpX - 0.5) : floor(tmpX + 0.5);
+    tmpX *= viewProvider.GridSize.getValue();
+
+    tmpY = tmpY / viewProvider.GridSize.getValue();
+    tmpY = tmpY < 0.0 ? ceil(tmpY - 0.5) : floor(tmpY + 0.5);
+    tmpY *= viewProvider.GridSize.getValue();
+
+    // Check if x within snap tolerance
+    if (x < tmpX + snapTol && x > tmpX - snapTol)
+        x = tmpX; // Snap X Mouse Position
+
+     // Check if y within snap tolerance
+    if (y < tmpY + snapTol && y > tmpY - snapTol)
+        y = tmpY; // Snap Y Mouse Position
+}
+
+void SnapManager::snapToLineMiddle(Base::Vector3d& pointToOverride, const Part::GeomLineSegment* line)
+{
+    Base::Vector3d startPoint = line->getStartPoint();
+    Base::Vector3d endPoint = line->getEndPoint();
+    Base::Vector3d midPoint = (startPoint + endPoint) / 2;
+
+    //Check if we are at middle of the line and if so snap to it.
+    if ((pointToOverride - midPoint).Length() < (endPoint - startPoint).Length() * 0.05)
+        pointToOverride = midPoint;
+}
+
+void SnapManager::snapToArcMiddle(Base::Vector3d& pointToOverride, const Part::GeomArcOfCircle* arc)
+{
+    Base::Vector3d centerPoint = arc->getCenter();
+    Base::Vector3d middleVec = (arc->getStartPoint() - centerPoint) + (arc->getEndPoint() - centerPoint);

```

Comment:
This formula seems incorrect or at least very weird.
Could you comment about this ?

Reply:
The formula to calculate middleVec? Or the whole function?
middleVec is just a vector that has the middle angle, but the length is not the radius if that is what you mean?
![image](https://user-images.githubusercontent.com/19984177/220085804-92dd7ce9-17b4-4c73-9df5-db25b30a726d.png)

Reply:
Worth a comment. Anyway, big problem for 180° arcs...

Reply:
I also found it strange. But I did not directly catch the issue with the 180 degrees.

If the vector is used to add it on the point, and the find the closest point to that on the curve, I think it should not be a problem, but I did not look into the specifics.

Reply:
Opps, middleVec.Length is zero in the implementation below and it is in the denominator (line 308...)

Reply:
> Opps, middleVec.Length is zero in the implementation below and it is in the denominator (line 308...)

:)

Reply:
Good catch. I haven't thought about this case.

Reply:
But still you marked this point as resolved. This is bit weird isn't it?

Reply:
Why? I solved it. Or did I missed something ?

Reply:
Oh yeah, I missed the commits

Reply:
It's because I cleaned the commit history so it went unoticed.